File: lib/utils/password_policy.dart
/// กฎความแข็งแรงของรหัสผ่าน แยกไว้ที่เดียวเพื่อให้หน้าอื่น เช่น หน้าเปลี่ยนรหัสผ่าน ใช้ซ้ำได้
class PasswordRule {
  const PasswordRule(this.label, this.passed);

  final String label;
  final bool passed;
}

class PasswordPolicy {
  PasswordPolicy._();

  static const int minLength = 8;

  /// ตรวจรหัสผ่านทีละข้อ ใช้ทั้งตอนกดสมัครและตอนแสดงรายการเงื่อนไขใต้ช่องกรอก
  static List<PasswordRule> check(
    String password, {
    String username = '',
    String email = '',
  }) {
    final lower = password.toLowerCase();
    final user = username.trim().toLowerCase();
    final emailName = email.split('@').first.trim().toLowerCase();
    final hasIdentity = (user.length >= 3 && lower.contains(user)) ||
        (emailName.length >= 3 && lower.contains(emailName));

    return [
      PasswordRule(
          'ยาวอย่างน้อย $minLength ตัวอักษร', password.length >= minLength),
      PasswordRule('มีตัวพิมพ์ใหญ่ A-Z อย่างน้อย 1 ตัว',
          RegExp(r'[A-Z]').hasMatch(password)),
      PasswordRule('มีตัวพิมพ์เล็ก a-z อย่างน้อย 1 ตัว',
          RegExp(r'[a-z]').hasMatch(password)),
      PasswordRule(
          'มีตัวเลข 0-9 อย่างน้อย 1 ตัว', RegExp(r'[0-9]').hasMatch(password)),
      PasswordRule('มีอักขระพิเศษ เช่น ! @ # \$ % อย่างน้อย 1 ตัว',
          RegExp(r'[^A-Za-z0-9]').hasMatch(password)),
      PasswordRule('ไม่มีช่องว่าง', !password.contains(' ')),
      PasswordRule('ไม่มีชื่อผู้ใช้หรือชื่ออีเมลอยู่ข้างใน', !hasIdentity),
    ];
  }

  /// คืนข้อความบอกเงื่อนไขข้อแรกที่ยังไม่ผ่าน ถ้าผ่านครบคืน null
  static String? firstError(
    String password, {
    String username = '',
    String email = '',
  }) {
    for (final rule in check(password, username: username, email: email)) {
      if (!rule.passed) return 'รหัสผ่านต้อง${rule.label}';
    }
    return null;
  }

  /// สัดส่วนเงื่อนไขที่ผ่าน 0.0 ถึง 1.0 เอาไว้ทำแถบวัดความแข็งแรง
  static double strength(
    String password, {
    String username = '',
    String email = '',
  }) {
    if (password.isEmpty) return 0;
    final rules = check(password, username: username, email: email);
    return rules.where((r) => r.passed).length / rules.length;
  }
}

File: backend/api/src/auth/password-policy.ts
// มิเรอร์กฎเดียวกับ lib/utils/password_policy.dart ฝั่ง Flutter ไว้ที่นี่
// เพราะ client-side validation ข้ามได้เสมอ (ยิง API ตรง ๆ ก็หลบกฎฝั่งแอปได้)
// ต้องบังคับซ้ำที่ server จึงจะถือว่าปลอดภัยจริง — ห้ามลบทิ้งแม้ Flutter จะเช็กแล้วก็ตาม
export function firstPasswordError(
  password: string,
  { username = '', email = '' }: { username?: string; email?: string },
): string | null {
  const rules: Array<[string, boolean]> = [
    ['ยาวอย่างน้อย 8 ตัวอักษร', password.length >= 8],
    ['มีตัวพิมพ์ใหญ่ A-Z อย่างน้อย 1 ตัว', /[A-Z]/.test(password)],
    ['มีตัวพิมพ์เล็ก a-z อย่างน้อย 1 ตัว', /[a-z]/.test(password)],
    ['มีตัวเลข 0-9 อย่างน้อย 1 ตัว', /[0-9]/.test(password)],
    ['มีอักขระพิเศษ เช่น ! @ # $ % อย่างน้อย 1 ตัว', /[^A-Za-z0-9]/.test(password)],
    ['ไม่มีช่องว่าง', !password.includes(' ')],
  ];

  const lower = password.toLowerCase();
  const user = username.trim().toLowerCase();
  const emailName = email.split('@')[0]?.trim().toLowerCase() ?? '';
  const hasIdentity =
    (user.length >= 3 && lower.includes(user)) ||
    (emailName.length >= 3 && lower.includes(emailName));
  rules.push(['ไม่มีชื่อผู้ใช้หรือชื่ออีเมลอยู่ข้างใน', !hasIdentity]);

  for (const [label, passed] of rules) {
    if (!passed) return `รหัสผ่านต้อง${label}`;
  }
  return null;
}
